### DISCUSSION TITLE
OptionButton improvements regarding per-item text and icons.

### DISCUSSION THREAD

**Discussion by Hatchet2k4** [godotengine/godot-proposals#9247] [GUI]
Using Godot 4.2.1. I am building a game that would allow players to select teams by color. I was hoping to use an OptionButton for this, and have each row's text match its color. Essentially turning this: 

![image](https://github.com/godotengine/godot-proposals/assets/5480348/dd333650-3635-4e78-b8f8-22377bcccbc5)

Into my mockup here:

![image](https://github.com/godotengine/godot-proposals/assets/5480348/998d2ccd-ee38-4a77-a3c5-f7a90d5f91e3)

Unfortunately there is no clear way to do this. I propose being able to have an option for each item's text to have individual themes/font properties that can be set, ie color, font, size, etc. 

I tried to work around this by creating icons for each team instead. However, I found that Godot didn't handle mixing rows that are just icons and just text very well:

![image](https://github.com/godotengine/godot-proposals/assets/5480348/2ca17018-ec1f-4eda-bd31-cf00a2ee592e)

I also tested with various icon sizes, essentially it is pushing the text to the right to match the width of the widest icon. If a row has text but no icon, I don't believe it should be pushed to the right as if it did have one. This could be resolved in a non-breaking way by having an option for icons to appear to the right of any text instead, so a row with text but no icon should still have text left aligned, and a row with an icon but no text would have the icon left aligned.

**Comment by Calinou** [comment]
> I tried to work around this by creating icons for each team instead. However, I found that Godot didn't handle mixing rows that are just icons and just text very well:

My suggestion here would be to use smaller icons with a square aspect ratio, and use a "crossed square" icon with a transparent background for the **None** option. See what Inkscape does in the bottom-left corner of its UI:

![image](https://github.com/godotengine/godot-proposals/assets/180032/95084a6e-0c1f-48f9-8551-aa212847a663)

Here's a mockup:

![image](https://github.com/godotengine/godot-proposals/assets/180032/e0dd9fa2-6fe4-4a21-a96b-9e060c588697)

**Reply by Hatchet2k4** [reply]
Yeah I ended up doing something like that. But I thought it was worth bringing up for discussion/potential improvements regardless.

**Comment by Mickeon** [comment]
> I propose being able to have an option for each item's text to have individual themes/font properties that can be set, ie color, font, size, etc.

While I would love this, I'd be wary because then one would have to question how far to take all of that modularity. Color, font, size... then one may then suggest styleboxes, underlines, checkbox texture overrides...

I know it's not pretty, but have you tried a more ad-hoc solution with CanvasItem's `draw_string()`?

**Reply by Hatchet2k4** [reply]
I would think it could just be limited to what is already supported in Themes. In any case I used icons with standard text. I suppose I could try using draw_string too.

**Comment by silent21century** [comment]
one easy way to achieve that is instead of using label's as items they could use a richtextlabel with BBCode enabled by default.

**Reply by Calinou** [reply]
OptionButton doesn't use Label nodes internally – it draws text using the CanvasItem `draw_string()` method. Reducing the number of internal nodes avoids overhead for OptionButtons that have lots of options.
